### PR TITLE
PHOENIX-5673 : The mutation state is silently getting cleared on the …

### PR DESCRIPTION
**Pull request by siddhimehta**
@ChinmaySKulkarni  @abhishek-chouhan Creating this PR to get feedback on the approach.
I have added a config to enable this behaviour ( default false).
Ran all the IT locally for the same

### CODE REVIEW

File: phoenix-core/src/main/java/org/apache/phoenix/jdbc/PhoenixStatement.java [hunk 1]
```diff
@@ -2166,4 +2169,26 @@ private void throwIfUnallowedUserDefinedFunctions(Map<String, UDFParseNode> udfP
...
+                    QueryServices.PENDING_MUTATIONS_DDL_THROW_ATTRIB,
+                    QueryServicesOptions.DEFAULT_PENDING_MUTATIONS_DDL_THROW);
+        if (stmt instanceof MutableStatement && !(stmt instanceof DMLStatement)
+                && state.getNumRows() > 0) {
+            if (throwUncommittedMutation) {
+                throw new SQLExceptionInfo.Builder(
+                        SQLExceptionCode.CANNOT_PERFORM_DDL_WITH_PENDING_MUTATIONS).build()
+                                .buildException();
+            } else {
+                LOGGER.info(
```
Comment (ChinmaySKulkarni) [inline phoenix-core/src/main/java/org/apache/phoenix/jdbc/PhoenixStatement.java]: Can we change this to `warn` level?

File: phoenix-core/src/test/java/org/apache/phoenix/execute/MutationStateTest.java [hunk 2]
```diff
@@ -175,4 +185,43 @@ public void testGetMutationBatchList() {
...
     }
 
+    @Test(expected = SQLException.class)
+    public void testPendingMutationsOnDDL() throws Exception {
+        Connection conn = null;
+        PhoenixConnection spy = null;
+        try {
+            Properties props = PropertiesUtil.deepCopy(TEST_PROPERTIES);
+            props.setProperty(QueryServices.PENDING_MUTATIONS_DDL_THROW_ATTRIB, "true");
+            conn = DriverManager.getConnection(getUrl(), props);
```
Comment (ChinmaySKulkarni) [inline phoenix-core/src/test/java/org/apache/phoenix/execute/MutationStateTest.java]: Surround connection with try-with-resources for auto-closing.

File: phoenix-core/src/test/java/org/apache/phoenix/execute/MutationStateTest.java [hunk 3]
```diff
@@ -175,4 +185,43 @@ public void testGetMutationBatchList() {
 
     }
 
+    @Test(expected = SQLException.class)
```
Comment (ChinmaySKulkarni) [inline phoenix-core/src/test/java/org/apache/phoenix/execute/MutationStateTest.java]: Same comment as above. Let's assert the exact sqlexception code too.

File: phoenix-core/src/test/java/org/apache/phoenix/execute/MutationStateTest.java [hunk 4]
```diff
@@ -175,4 +185,43 @@ public void testGetMutationBatchList() {
...
+            }
+        }
+    }
+
+    @Test
+    public void testNoPendingMutationsOnDDL() throws Exception {
+        Connection conn = null;
+        try {
+            conn = DriverManager.getConnection(getUrl());
+            conn.createStatement().execute("create table MUTATION_TEST1"
```
Comment (ChinmaySKulkarni) [inline phoenix-core/src/test/java/org/apache/phoenix/execute/MutationStateTest.java]: Same comment about adding try-with-resources for the connection

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/MutationStateIT.java [hunk 5]
```diff
@@ -499,4 +499,27 @@ public void testSplitMutationsIntoSameGroupForSingleRow() throws Exception {
...
+        PreparedStatement stmt =
+                conn.prepareStatement("UPSERT INTO " + tableName
+                        + " (inst,host,\"DATE\") VALUES(?,'b',CURRENT_DATE())");
+        stmt.setString(1, "a");
+        stmt.execute();
+
+        // Create a ddl statement
+        String tableName2 = generateUniqueName();
+        String ddl = "CREATE TABLE " + tableName2 + " (V BIGINT PRIMARY KEY, K BIGINT)";
+        conn.createStatement().execute(ddl);
```
[the remark this answers is not in the document]
Comment (siddhimehta) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/MutationStateIT.java]: Done. Used ExpectedException which asserts exception & message and fails if no exceptions

File: phoenix-core/src/test/java/org/apache/phoenix/execute/MutationStateTest.java (hunk 3 above)
Comment (siddhimehta) [reply phoenix-core/src/test/java/org/apache/phoenix/execute/MutationStateTest.java]: Same as above. Used ExcpetedExceptionRule

File: phoenix-core/src/test/java/org/apache/phoenix/execute/MutationStateTest.java (hunk 2 above)
Comment (siddhimehta) [reply phoenix-core/src/test/java/org/apache/phoenix/execute/MutationStateTest.java]: done

**Comment by siddhimehta** [comment]
@ChinmaySKulkarni Tested view creation and index creation also

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/MutationStateIT.java [hunk 6]
```diff
@@ -499,4 +501,32 @@ public void testSplitMutationsIntoSameGroupForSingleRow() throws Exception {
...
+    @Rule
+    public ExpectedException exceptionRule = ExpectedException.none();
+
+    @Test
+    public void testDDLwithPendingMutations() throws Exception {
+        String tableName = generateUniqueName();
+        ensureTableCreated(getUrl(), tableName, TestUtil.PTSDB_NAME, null, null, null);
+        Properties props = new Properties();
+        props.setProperty(QueryServices.PENDING_MUTATIONS_DDL_THROW_ATTRIB, Boolean.toString(true));
+        Connection conn = DriverManager.getConnection(getUrl(), props);
```
Comment (ChinmaySKulkarni) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/MutationStateIT.java]: nit: try-with-resources to auto-close connection and statement.

**Pull request closed without merging**